SharedLibs/TileMapHelper/src/model/TileMapModel.ts

```diff
@@ -352,25 +504,243 @@ export class EditableTileMapLayer extends AbstractEditableLayer {
     this._alpha = alpha;
   }
 
+  isEmpty(): boolean {
+    return this._tiles.every((row) => row.every((cell) => cell === 0));
+  }
+
+  reduceDimensions(
+    columnsToPop: number,
+    columnsToShift: number,
+    rowsToPop: number,
+    rowsToShift: number
+  ) {
+    const initialRowCount = this._tiles.length;
+    const initialColumnCount = this._tiles[0].length;
+
+    if (rowsToPop > 0 || rowsToShift > 0) {
+      this._tiles = this._tiles.slice(
+        rowsToShift,
+        rowsToPop ? -rowsToPop : undefined
+      );
+    }
+    if (columnsToPop > 0 || columnsToShift > 0) {
+      this._tiles.forEach((row, rowIndex) => {
+        this._tiles[rowIndex] = this._tiles[rowIndex].slice(
+          columnsToShift,
+          columnsToPop ? -columnsToPop : undefined
+        );
+      });
+    }
+    // TODO: Instead of setting the dimensions directly, should it call a method on
+    // EditableTileMap that will iterates over all the layers to change their dimensions?
+    this.tileMap.setDimensionX(
+      initialColumnCount - columnsToPop - columnsToShift
+    );
+    this.tileMap.setDimensionY(initialRowCount - rowsToPop - rowsToShift);
+  }
+
+  increaseDimensions(
+    columnsToAppend: number,
+    columnsToUnshift: number,
+    rowsToAppend: number,
+    rowsToUnshift: number
+  ) {
+    const initialRowCount = this._tiles.length;
+    const initialColumnCount = this._tiles[0].length;
+    if (columnsToAppend > 0 || columnsToUnshift > 0) {
+      this._tiles.forEach((row, rowIndex) => {
+        const newRow = new Int32Array(
+          initialColumnCount + columnsToAppend + columnsToUnshift
+        ).fill(0);
+        newRow.set(row, columnsToUnshift);
+        this._tiles[rowIndex] = newRow;
+      });
+    }
+    if (rowsToAppend > 0 || rowsToUnshift > 0) {
+      this._tiles.unshift(
+        ...new Array(rowsToUnshift)
+          .fill(0)
+          .map(() =>
+            new Int32Array(
+              initialColumnCount + columnsToAppend + columnsToUnshift
+            ).fill(0)
+          )
+      );
+
+      this._tiles.length = initialRowCount + rowsToAppend + rowsToUnshift;
+
+      for (
+        let rowIndex = initialRowCount + rowsToUnshift;
+        rowIndex < this._tiles.length;
+        rowIndex++
+      ) {
+        this._tiles[rowIndex] = new Int32Array(
+          initialColumnCount + columnsToAppend + columnsToUnshift
+        ).fill(0);
+      }
+    }
+    // TODO: Instead of setting the dimensions directly, should it call a method on
+    // EditableTileMap that will iterates over all the layers to change their dimensions?
+    this.tileMap.setDimensionX(
+      initialColumnCount + columnsToAppend + columnsToUnshift
+    );
+    this.tileMap.setDimensionY(initialRowCount + rowsToAppend + rowsToUnshift);
```

Comment:
I think that this method `EditableTileMapLayer.increaseDimensions` should be private. Something like `EditableTileMap.setBounds` should indeed ask every layer to change of dimension (even if the origin is snapped at the top left with a todo to handle custom origin later as it's important to avoid column and row index to change as the map expand which will feel odd if users make a loop).

Reply:
As seen together, this is used in the editor only. Do you think I can leave it for later?

Reply:
I moved the responsibility to change tilemap dimensions to the tilemap itself.